### PR TITLE
OHRM5X-102: Develop Admin - Qualification - Licenses - FE & BE API

### PR DESCRIPTION
**Pull request by ireshe**

### CODE REVIEW

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationLicense/License.vue [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+            edit: {
+              onClick: this.onClickEdit,
+              props: {
+                name: 'pencil-fill',
+              },
+            },
+          },
+        },
+      ],
+      editItem: null,
```
Comment (RajithaKumara) [inline symfony/client/src/orangehrmAdminPlugin/pages/qualificationLicense/License.vue:110]: @Super-Chama do we need this `editItem`?

File: symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+use OrangeHRM\Core\Api\V2\Model\ArrayModel;
+use OrangeHRM\Core\Api\V2\ParameterBag;
+use OrangeHRM\Core\Api\V2\RequestParams;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointCreateResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointDeleteResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointGetAllResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointGetOneResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointUpdateResult;
+use OrangeHRM\Entity\License;
+use Orangehrm\Rest\Api\Exception\RecordNotFoundException;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php:38]: `use OrangeHRM\Core\Api\V2\Exception\RecordNotFoundException;`

File: symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+     * @throws DaoException
+     * @throws Exception
+     */
+    public function getOne(): EndpointGetOneResult
+    {
+        // TODO:: Check data group permission
+        $id = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, self::PARAMETER_ID);
+        $license = $this->getLicenseService()->getLicenseById($id);
+        if (!$license instanceof License) {
+            throw new RecordNotFoundException('No Record Found');
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php:83]: `throw new RecordNotFoundException();`

File: symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+     */
+    public function getAll(): EndpointGetAllResult
+    {
+        // TODO:: Check data group permission
+
+        $licenseParamHolder = new LicenseSearchFilterParams();
+        $this->setSortingAndPaginationParams($licenseParamHolder);
+        $licenses = $this->getLicenseService()->getLicenseList($licenseParamHolder);
+        $count = $this->getLicenseService()->getLicenseCount($licenseParamHolder);
+        return new EndpointGetAllResult(LicenseModel::class, $licenses, new ParameterBag(['total' => $count]));
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php:102]: `new ParameterBag([CommonParams::PARAMETER_TOTAL => $count])`

File: symfony/plugins/orangehrmAdminPlugin/Controller/QualificationLicenseController.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ * Customer shall remain vested in OrangeHRM Inc. Any rights not expressly granted herein are
+ * reserved to OrangeHRM Inc.
+ *
+ * Please refer http://www.orangehrm.com/Files/OrangeHRM_Commercial_License.pdf for the license which includes terms and conditions on using this software.
+ *
+ */
+
+
+namespace OrangeHRM\Admin\Controller;
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Controller/QualificationLicenseController.php:23]: multiple blank lines. one enough

File: symfony/plugins/orangehrmAdminPlugin/Controller/SaveQualificationLicenseController.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * vested in OrangeHRM Inc. For the avoidance of doubt title and all intellectual property
+ * rights to any design, new software, new protocol, new interface, enhancement, update,
+ * derivative works, revised screen text or any other items that OrangeHRM Inc creates for
+ * Customer shall remain vested in OrangeHRM Inc. Any rights not expressly granted herein are
+ * reserved to OrangeHRM Inc.
+ *
+ * Please refer http://www.orangehrm.com/Files/OrangeHRM_Commercial_License.pdf for the license which includes terms and conditions on using this software.
+ *
+ */
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Controller/SaveQualificationLicenseController.php:20]: 2 blank lines. remove one

File: symfony/plugins/orangehrmAdminPlugin/Controller/SaveQualificationLicenseController.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        $id = $request->get('id');
+        if ($id) {
+            $component = new Component('license-edit');
+            $component->addProp(new Prop('license-id', Prop::TYPE_NUMBER, $id));
+        } else {
+            $component = new Component('license-save');
+        }
+        $this->setComponent($component);
+    }
+}
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Controller/SaveQualificationLicenseController.php:43]: add new line end of the file. configure the IDE

File: symfony/plugins/orangehrmAdminPlugin/Dao/LicenseDao.php [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        }
+    }
+
+    /**
+     * @param LicenseSearchFilterParams $licenseSearchParamHolder
+     * @return Paginator
+     */
+    public function getLicenseListPaginator(LicenseSearchFilterParams $licenseSearchParamHolder): Paginator
+    {
+        $q = Doctrine::getEntityManager()->getRepository(License::class)->createQueryBuilder('l');
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/LicenseDao.php:105]: use `$this->createQueryBuilder(...)`

File: symfony/plugins/orangehrmAdminPlugin/Dao/LicenseDao.php [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+     * @return int
+     * @throws DaoException
+     */
+    public function deleteLicenses(array $toDeleteIds): int
+    {
+        try {
+            $q = $this->createQueryBuilder(License::class, 'e');
+            $q->delete(License::class, 'l')
+                ->set('l.deleted', true)
+                ->where($q->expr()->in('l.id', $toDeleteIds));
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/LicenseDao.php:142]:
```php
->where($q->expr()->in('l.id', ':ids'));
->setParameter('ids', $toDeleteIds);
```

File: symfony/plugins/orangehrmAdminPlugin/Dao/LicenseDao.php [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+    /**
+     * @param $licenseName
+     * @return bool
+     * @throws DaoException
+     */
+    public function isExistingLicenseName($licenseName): bool
+    {
+        try {
+            $q = $this->getRepository(License::class)->createQueryBuilder('l');
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/LicenseDao.php:157]: use `$this->createQueryBuilder(...)`

File: symfony/plugins/orangehrmAdminPlugin/Dao/LicenseDao.php [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        }
+        return new Paginator($q);
+    }
+
+    /**
+     * @param LicenseSearchFilterParams $licenseSearchParamHolder
+     * @return int
+     * @throws DaoException
+     */
+    public function getLicenseCount(LicenseSearchFilterParams $licenseSearchParamHolder): int
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/LicenseDao.php:121]: `$licenseSearchParamHolder` => `$licenseSearchFilterParams`

File: symfony/plugins/orangehrmAdminPlugin/Dto/LicenseSearchFilterParams.php [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Admin\Dto;
+
+use OrangeHRM\Core\Dto\FilterParams;
+
+class LicenseSearchFilterParams extends FilterParams
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dto/LicenseSearchFilterParams.php:24]: all class properties are unnecessary. make this class as;
```php
class LicenseSearchFilterParams extends FilterParams
{
}
```

File: symfony/plugins/orangehrmAdminPlugin/Service/LicenseService.php [hunk 13]
```diff
@@ -17,84 +17,116 @@
...
         return $this->licenseDao;
     }
 
     /**
-     * @ignore
+     * @param $licenseDao
+     * @return void
      */
-    public function setLicenseDao($licenseDao) {
+    public function setLicenseDao($licenseDao)
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Service/LicenseService.php:52]: `public function setLicenseDao(LicenseDao $licenseDao):void`

File: symfony/plugins/orangehrmAdminPlugin/Service/LicenseService.php [hunk 14]
```diff
@@ -17,84 +17,116 @@
...
-     * @param License $license 
+     *
+     * @param License $license
      * @return NULL Doesn't return a value
+     * @throws \DaoException
+     * @version 2.6.12
      */
-    public function saveLicense(License $license) {        
-        $this->getLicenseDao()->saveLicense($license);        
+    public function saveLicense(License $license)
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Service/LicenseService.php:67]: return type. may be you can return `$this->getLicenseDao()->saveLicense($license);`

File: symfony/plugins/orangehrmAdminPlugin/entity/License.php [hunk 15]
```diff
@@ -0,0 +1,85 @@
+<?php
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/entity/License.php:2]: add copyright header

File: symfony/plugins/orangehrmAdminPlugin/entity/License.php [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    /**
+     * @var Collection
+     *
+     * @ORM\ManyToMany(targetEntity="EmployeeLicense", mappedBy="License")
+     * @ORM\JoinColumns({
+     *   @ORM\JoinColumn(name="id", referencedColumnName="license_id")
+     * })
+     */
+    private $EmployeeLicense;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/entity/License.php:43]: remove this bidirectional mapping & the constructor. once remove test properly

File: symfony/plugins/orangehrmAdminPlugin/Service/LicenseService.php [hunk 17]
```diff
@@ -17,84 +17,116 @@
...
      * Deletes licenses
-     * 
-     * @version 2.6.12 
+     *
      * @param array $toDeleteIds An array of IDs to be deleted
      * @return int Number of records deleted
-     */    
-    public function deleteLicenses($toDeleteIds) {
+     * @throws DaoException
+     * @version 2.6.12
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Service/LicenseService.php:126]: remove `@version` tag in all doc comments

File: dbscript/dbscript-2.sql [hunk 18]
```diff
@@ -3559,7 +3559,7 @@ UPDATE `ohrm_menu_item` SET `status` = '0' WHERE `ohrm_menu_item`.`menu_title` =
 UPDATE `ohrm_menu_item` SET `status` = '0' WHERE `ohrm_menu_item`.`menu_title` = 'Locations';
 UPDATE `ohrm_menu_item` SET `status` = '0' WHERE `ohrm_menu_item`.`menu_title` = 'Skills';
 UPDATE `ohrm_menu_item` SET `status` = '0' WHERE `ohrm_menu_item`.`menu_title` = 'Education';
-UPDATE `ohrm_menu_item` SET `status` = '0' WHERE `ohrm_menu_item`.`menu_title` = 'Licenses';
+UPDATE `ohrm_menu_item` SET `status` = '1' WHERE `ohrm_menu_item`.`menu_title` = 'license';
```
Comment (RajithaKumara) [inline dbscript/dbscript-2.sql:3562]: remove this line

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+     * @return License
+     * @throws RecordNotFoundException
+     * @throws \OrangeHRM\Core\Exception\DaoException
+     */
+    public function saveLicense(): License
+    {
+        $id = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, self::PARAMETER_ID);
+        $name = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_NAME);
+        if (!empty($id)) {
+            $education = $this->getLicenseService()->getLicenseById($id);
```
Comment (KaveeshaSilva1996) [inline symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php:141]: Is $education variable name correct?

Comment (ireshe) [reply symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php:141]: @KaveeshaSilva1996 no actually It was from the education API i created

File: symfony/plugins/orangehrmAdminPlugin/Dto/LicenseSearchFilterParams.php (hunk 12 above)
Comment (ireshe) [reply symfony/plugins/orangehrmAdminPlugin/Dto/LicenseSearchFilterParams.php:24]: added these

`public const ALLOWED_SORT_FIELDS = ['l.name'];

    public function __construct()
    {
        $this->setSortField('l.name');
    }`

File: symfony/plugins/orangehrmAdminPlugin/entity/License.php (hunk 16 above)
Comment (ireshe) [reply symfony/plugins/orangehrmAdminPlugin/entity/License.php]: done, tested using postman and passed all test cases

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationLicense/SaveLicense.vue [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+          return (!!v && v.trim() !== '') || 'Required';
+        });
+        this.rules.name.push(v => {
+          return (v && v.length <= 50) || 'Should not exceed 50 characters';
+        });
+        this.rules.name.push(v => {
+          const index = data.findIndex(item => item.name === v);
+          return index === -1 || 'License name should be unique';
+        });
+        this.isLoading = false;
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmAdminPlugin/pages/qualificationLicense/SaveLicense.vue:122]: remove `this.isLoading = false;`

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationLicense/License.vue [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+            });
+          })
+          .then(() => {
+            this.isLoading = false;
+            this.resetDataTable();
+          });
+      }
+    },
+    async resetDataTable() {
+      this.$refs.dTable.checkedItems = [];
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmAdminPlugin/pages/qualificationLicense/License.vue:204]: change to `this.checkedItems = [];`

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationLicense/License.vue [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+              @click="onClickDeleteSelected"
+              class="orangehrm-horizontal-margin"
+            />
+          </div>
+          <oxd-text tag="span" v-else>{{ itemsCountText }}</oxd-text>
+        </div>
+      </div>
+      <div class="orangehrm-container">
+        <oxd-card-table
+          ref="dTable"
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmAdminPlugin/pages/qualificationLicense/License.vue:55]: remove `ref="dTable"`
add `:loading="isLoading"`

File: symfony/plugins/orangehrmAdminPlugin/entity/License.php [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * License
+ *
+ * @ORM\Table(name="ohrm_license")
+ * @ORM\Entity
+ */
+class License
+{
+    public const NO_OF_RECORDS_PER_PAGE = 50;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/entity/License.php:34]: if you are not using 3 costs in ur implementation, we can remove them

File: symfony/plugins/orangehrmAdminPlugin/test/Dao/LicenseDaoTest.php [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+    protected $fixture;
+
+    /**
+     * Set up method
+     * @throws Exception
+     */
+    protected function setUp(): void
+    {
+        $this->licenseDao = new LicenseDao();
+        $this->fixture = Config::get('ohrm_plugins_dir') . '/orangehrmAdminPlugin/test/fixtures/LicenseDao.yml';
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Dao/LicenseDaoTest.php:46]: can you replace `Config::get('ohrm_plugins_dir')` to `Config::get(Config::PLUGINS_DIR)`

**Pull request merged**